File: dream-server/src/dream_server/engine/preprocessing.py
"""Model-specific image preprocessing and deprocessing."""

from __future__ import annotations

from typing import Literal

import torch
from torchvision import transforms

PreprocessingMode = Literal["imagenet", "inception", "none"]

# TF Keras convention for MobileNetV2/InceptionV3: [-1, 1]
INCEPTION_TRANSFORM = transforms.Compose([
    transforms.ToTensor(),  # [0, 255] → [0, 1]
    transforms.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5]),  # → [-1, 1]
])

# PyTorch/ImageNet convention for VGG19
IMAGENET_TRANSFORM = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])


def get_transform(mode: PreprocessingMode) -> transforms.Compose:
    if mode == "inception":
        return INCEPTION_TRANSFORM
    elif mode == "imagenet":
        return IMAGENET_TRANSFORM
    else:
        return transforms.Compose([transforms.ToTensor()])


def deprocess(tensor: torch.Tensor, mode: PreprocessingMode) -> torch.Tensor:
    """Reverse preprocessing to get [0, 1] range tensor for display."""
    if mode == "inception":
        return tensor.mul(0.5).add(0.5).clamp(0, 1)
    elif mode == "imagenet":
        mean = torch.tensor([0.485, 0.456, 0.406], device=tensor.device).view(3, 1, 1)
        std = torch.tensor([0.229, 0.224, 0.225], device=tensor.device).view(3, 1, 1)
        return tensor.mul(std).add(mean).clamp(0, 1)
    else:
        return tensor.clamp(0, 1)

File: src/dream/engine/preprocessing.ts
import * as tf from "@tensorflow/tfjs";

export type PreprocessingMode = "imagenet" | "inception" | "none";

/**
 * Preprocess an image tensor for model input.
 *
 * - "inception": scale pixels from [0, 255] to [-1, 1]  (TF Keras MobileNetV2, InceptionV3)
 * - "imagenet": subtract ImageNet mean, divide by std   (PyTorch convention, VGG19)
 * - "none": pass through as-is
 */
export function preprocess(
  image: tf.Tensor3D,
  mode: PreprocessingMode,
): tf.Tensor3D {
  return tf.tidy(() => {
    switch (mode) {
      case "inception":
        // [0, 255] → [-1, 1]
        return image.div(127.5).sub(1) as tf.Tensor3D;

      case "imagenet": {
        // [0, 255] → [0, 1] → subtract mean → divide by std
        const normalized = image.div(255);
        const mean = tf.tensor1d([0.485, 0.456, 0.406]);
        const std = tf.tensor1d([0.229, 0.224, 0.225]);
        return normalized.sub(mean).div(std) as tf.Tensor3D;
      }

      case "none":
        return image;
    }
  });
}

/**
 * Reverse preprocessing to [0, 1] float range for tf.browser.toPixels().
 */
export function deprocess(
  image: tf.Tensor3D,
  mode: PreprocessingMode,
): tf.Tensor3D {
  return tf.tidy(() => {
    switch (mode) {
      case "inception":
        // [-1, 1] → [0, 1]
        return image.add(1).mul(0.5).clipByValue(0, 1) as tf.Tensor3D;

      case "imagenet": {
        // Undo normalization → [0, 1]
        const mean = tf.tensor1d([0.485, 0.456, 0.406]);
        const std = tf.tensor1d([0.229, 0.224, 0.225]);
        return image
          .mul(std)
          .add(mean)
          .clipByValue(0, 1) as tf.Tensor3D;
      }

      case "none":
        // Assume [0, 255], scale to [0, 1]
        return image.div(255).clipByValue(0, 1) as tf.Tensor3D;
    }
  });
}
